### PR TITLE
[Diagnostics] handle ‘open’ access control; introduce %error modifier

### PR DESCRIPTION
**Pull request by jtbandes**
I noticed that `open` was missing from most of the access control–related diagnostics. This PR also adds an assert to the `%select`-parsing code, so hopefully we can catch these kinds of errors earlier in the future.

cc @jrose-apple who worked on `fileprivate`, and @rjmccall who implemented `open`.

@practicalswift – maybe this is another candidate for fuzzing :)

### CODE REVIEW

**Comment by jtbandes** [comment]
@swift-ci please smoke test

**Comment by jtbandes** [comment]
@swift-ci please test

**Comment by jtbandes** [comment]
@swift-ci please test Linux platform

**Comment by jtbandes** [comment]
I don't really understand the errors coming from Linux. They don't seem to be related to my changes, but I'm not certain. @shahmishal, who should I ask about these?

**Comment by jtbandes** [comment]
@swift-ci please smoke test

**Comment by jtbandes** [comment]
@swift-ci please test Linux platform

**Comment by practicalswift** [comment]
@jtbandes I'm always open to new fuzzing opportunities! What would be the best way to fuzz this? :-)

**Comment by jtbandes** [comment]
The access control diagnostics, in particular, are usually related to discrepancies in the accessibility on a decl vs. its use site, or a decl vs. one of its children, etc. You could try generating random combinations of those (or, perhaps you could just generate **all** of them, and verify that you've produced all possible diagnostics by cross-referencing `DiagnosticsSema.def`).

As for actually detecting the errors, it seems to be a common idiom to use `ALL CAPS` in diagnostics when an unexpected situation has occurred. In some cases you might also find two spaces in a row where a word is missing. (My PR addresses the latter by adding an `assert` to ensure a string has been provided. We could also consider introducing something like `%error` that would explicitly indicate an error and trigger an assertion, e.g. `%select{fileprivate|private|internal|public|%error}`.)

**Comment by CodaFi** [comment]
@swift-ci please test Linux platform.

**Comment by jrose-apple** [comment]
We would *definitely* need tests for all these diagnostics before you merge this. :-)

Many of the diagnostics will never actually generate the `open` case because they're talking about use sites, not override sites, and therefore never actually generate "open" as the level of access. (In some sense, "open is a kind of public" for these diagnostics.) That's why we haven't needed this yet. But in other cases it does make things clearer.

We've actually encountered some of the bad diagnostics in real life, usually because violating access control in one place breaks assumptions elsewhere. So I'd be hesitant to replace the bad cases with crashes, at least in release builds. (We already have problems with this, actually—parts of the compiler will assert that access control makes sense even when it doesn't.)

**Comment by jtbandes** [comment]
Yeah, my motivation for this was when I noticed a case where the `%select{...}` was expanded to an empty string, because something was `open` but values were only given for up to `public`. Some of them already indicate truly unexpected situations with ALL CAPS; I figured it might be better to make that a real assertion, since it really shouldn't be happening.

**Comment by jrose-apple** [comment]
Ah, I'm pretty sure the diagnostic engine is already supposed to assert when it reaches past the end of a select…

**Comment by shahmishal** [comment]
@tfiala Can you please help us with these LLDB test are failures? Thanks!

https://ci.swift.org/job/swift-PR-Linux/4474/
```
12:24:30 =============
12:24:30 Issue Details
12:24:30 =============
12:24:30 ERROR: testREPL (repl/structs/TestREPLStructs.py)
12:24:30 ERROR: testREPL (repl/type_lookup/TestREPLTypeLookup.py)
12:24:30 ERROR: testREPL (repl/pounwrapping/TestPOUnwrapping.py)
12:24:30 ERROR: testREPL (repl/classes/TestREPLClasses.py)
12:24:30 ERROR: testREPL (repl/lazy_reverse/TestLazyReverse.py)
12:24:30 ERROR: testREPL (repl/recursive_class/TestREPLRecursiveClass.py)
```

**Comment by jtbandes** [comment]
@jrose-apple Yeah, I thought that would be a good idea, which is why I added it in this PR :)

File: include/swift/AST/DiagnosticsSema.def
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: Now that `open` is the max, this `PUBLIC` should be `public`.

**Review by CodaFi — CHANGES_REQUESTED** [review]

File: include/swift/AST/DiagnosticsSema.def
Comment (jtbandes) [inline include/swift/AST/DiagnosticsSema.def]: I'm not totally clear on whether these should be allowed:

```
public class Foo {}

public extension Foo {
    open func bar() {}  // currently warns
}

extension Foo {
    open func bar() {}  // currently doesn't warn
}
```

**Comment by jtbandes** [comment]
After a long hiatus, I've updated this PR to replace the `ALLCAPS` diag convention with an explicit `%error`, which triggers an `assert()` in debug builds. I'm also still `assert()`ing when an argument goes past the bounds of a `%select`.

@practicalswift as mentioned in an earlier comment, if you're interested, it could be nice to fuzz / generate all combinations of access levels with types, typealiases, vars, lets, inheritance, protocols, etc. to check that the access control diagnostics don't hit any of these `%error` cases.

@swift-ci please smoke test

**Comment by jtbandes** [comment]
@swift-ci please smoke test

**Review by CodaFi — APPROVED** [review]

**Comment by jtbandes** [comment]
also cc stakeholder @DougGregor :)

**Comment by jtbandes** [comment]
Made a minor update to avoid crashing if `%error` is encountered in a release build. It will now output `<<ERROR>>`.

@swift-ci please smoke test

**Comment by jrose-apple** [comment]
:-( I'm calling this a failure of the new reviewing feature of GitHub. I had outstanding comments, which @jtbandes addressed but which I hadn't seen yet. @CodaFi said it was good from his end, and @slavapestov took that and the tests passing as enough to merge. I think it would have been fair to wait for me to sign off on this as well. What should we do differently next time?

**Comment by jtbandes** [comment]
Hmm. Sorry about that. By adding you as a reviewer I had intended for you to get another look at it, although github definitely doesn't enforce that... I could've also made it explicit in a comment. Maybe a policy that something shouldn't be merged until all the reviewers have approved?

**Comment by jtbandes** [comment]
Also, would it have helped for me to @-mention you in a comment? I'm not sure what kind of notifications github sends when adding reviewers...

**Comment by jtbandes** [comment]
Another solution would be for you to "request changes", which is a marker that would stick around until you re-review/approve. I'm not sure what the settings in this repo are, but it can be configured to prevent merging until the review is approved (including no changes requested by any reviewer).

**Comment by jrose-apple** [comment]
Yeah. I'm hesitant to use "request changes" because it feels harsh (big red X), but it certainly would have made my desire clearer here.

**Comment by jtbandes** [comment]
I won't take it personally :)

**Comment by jtbandes** [comment]
@jrose-apple Was there anything you still wanted changed? I'd be happy to make a PR to refine further if you have more ideas.

**Comment by jrose-apple** [comment]
No, I think I would have OKed it anyway. (Thank you as always for making things better!) I'm just a bit nervous of the around-the-edges access control problems—if you break the rules in one place and then use that declaration elsewhere I think we might hit some of these error clauses.

**Comment by jtbandes** [comment]
That might be true. I think it'll be helpful to try and programmatically generate a bunch of test inputs.

**Pull request merged**
